### PR TITLE
[SYCL][ESIMD][NewPM] Add ESIMDVerifierPass to new PM pipeline in CodeGen

### PR DESCRIPTION
**Pull request by mlychkov**
Signed-off-by: Mikhail Lychkov <mikhail.lychkov@intel.com>

### CODE REVIEW

**Review by AlexeySachkov — APPROVED** [review]

**Review by smanna12 — APPROVED** [review]

**Review by elizabethandrews — APPROVED** [review]

**Comment by mlychkov** [comment]
The full set of fixes for new PM also validated in #4860.

File: clang/lib/CodeGen/BackendUtil.cpp [hunk 1]
```diff
@@ -1385,6 +1385,12 @@ void EmitAssemblyHelper::RunOptimizationPipeline(
     // configure the pipeline.
     OptimizationLevel Level = mapToLevel(CodeGenOpts);
 
+    if (LangOpts.SYCLIsDevice)
```
Comment (kbobrovs) [inline clang/lib/CodeGen/BackendUtil.cpp:1388]: ESIMD verification should ideally run even if LLVM passes are disabled, as this is logically part of FE semantic analysis.
@sndmitriev - what do you think?

Comment (AlexeySachkov) [reply clang/lib/CodeGen/BackendUtil.cpp:1388]: > ESIMD verification should ideally run even if LLVM passes are disabled, as this is logically part of FE semantic analysis.
@sndmitriev - what do you think?

Please note that `disable-llvm-passes` disables even such passes as AlwaysInline, which are required to preserve the semantics of `always_inline` attribute, i.e. the flag literally says disable `llvm` passes, not optimization or transformation passes, but `llvm` passes.

My understanding is that the flag is not intended to be used by end users unless they know what they are doing and therefore, I would prefer to leave ESIMD verification disabled if `-disable-llvm-passes` is present

Comment (bader) [reply clang/lib/CodeGen/BackendUtil.cpp:1388]: I agree with Alexey. `-disable-llvm-passes` is supposed to disable all LLVM passes with no exceptions.
I think what Konst has in mind is that ESIMD verifier pass should run even with `-O0`.

Comment (bader) [reply clang/lib/CodeGen/BackendUtil.cpp:1388]: > ESIMD verification should ideally run even if LLVM passes are disabled, as this is logically part of FE semantic analysis.

Can we move the logic of this pass to FE Sema then?

Comment (mlychkov) [reply clang/lib/CodeGen/BackendUtil.cpp:1388]: @kbobrovs @sndmitriev Could you please provide any comments?

Comment (kbobrovs) [reply clang/lib/CodeGen/BackendUtil.cpp:1388]: Sorry for delay.
Unfortunately, we can't move this to Sema as it was identified as very time-consuming task by the FE team, so we went with the simpler solution.

> I think what Konst has in mind is that ESIMD verifier pass should run even with -O0.

Yes, exactly.

**Review by kbobrovs — CHANGES_REQUESTED** [review]
Not running ESIMDVerifier with O0 is a deal breaker for ESIMD.

**Comment by AlexeySachkov** [comment]
> Not running ESIMDVerifier with O0 is a deal breaker for ESIMD.

This patch doesn't disable `ESIMDVerifier` in `-O0`, because `-disable-llvm-passes` is not `-O0`. At `-O0` some passes are still run

**Comment by mlychkov** [comment]
> Not running ESIMDVerifier with O0 is a deal breaker for ESIMD.

The test is updated to check that ESIMDVerifier pass is executed for O0 opt level as well.

**Review by kbobrovs — APPROVED** [review]

**Comment by mlychkov** [comment]
Looks like failure of HIP tests is caused by https://github.com/intel/llvm/pull/5374.

**Comment by bader** [comment]
> Looks like failure of HIP tests is caused by #5374.

I don't think so. These tests were passing in #5374 [pre-commit](https://github.com/intel/llvm/runs/4939081347?check_suite_focus=true). It might be due to a conflict with some other changes or maybe tested sources didn't have #5374 changes. Could you update your branch to make sure that #5374 is included, please?

**Review by smanna12 — APPROVED** [review]

**Comment by mlychkov** [comment]
Another one failure due to CI infra problems. Raised an issue and restarted failed job.

**Pull request merged**